backend/src/util/http_reader.rs

```diff
@@ -167,51 +169,88 @@ impl AsyncRead for HttpReader {
+                        Some(Err(StdIOError::new(
+                            std::io::ErrorKind::InvalidInput,
+                            format!("more bytes returned than expected"),
+                        )))
+                    } else {
+                        buf.put_slice(&*bytes);
+                        Some(Ok(bytes.len()))
+                    }
+                }
+                Some(Err(e)) => Some(Err(StdIOError::new(std::io::ErrorKind::Interrupted, e))),
+                None => None,
+            })
+        }
         let this = self.project();
 
-        let mut fut = if let Some(fut) = this.read_in_progress.take() {
-            fut
-        } else {
-            HttpReader::get_range(
-                *this.range_unit,
-                this.http_client.clone(),
-                this.http_url.clone(),
-                *this.cursor_pos,
-                buf.remaining(),
-                *this.total_bytes,
-            )
-            .boxed()
-        };
-
-        let res_poll = fut.as_mut().poll(cx);
-        trace!("Polled with remaining bytes in buf: {}", buf.remaining());
-
-        match res_poll {
-            Poll::Ready(result) => match result {
-                Ok(data_chunk) => {
-                    trace!("data chunk: len: {}", data_chunk.len());
-                    trace!("buf filled len: {}", buf.filled().len());
-
-                    if data_chunk.len() <= buf.remaining() {
-                        buf.put_slice(&data_chunk);
-                        *this.cursor_pos += data_chunk.len();
-
-                        Poll::Ready(Ok(()))
-                    } else {
-                        buf.put_slice(&data_chunk);
+        loop {
+            let mut in_progress = match this.read_in_progress.take() {
+                ReadInProgress::Complete(mut body) => match poll_complete(&mut body, cx, buf) {
+                    Poll::Pending => {
+                        *this.read_in_progress = ReadInProgress::Complete(body);
+                        return Poll::Pending;
+                    }
+                    Poll::Ready(Some(Ok(len))) => {
+                        *this.read_in_progress = ReadInProgress::Complete(body);
+                        *this.cursor_pos += len;
 
-                        Poll::Ready(Ok(()))
+                        return Poll::Ready(Ok(()));
+                    }
+                    Poll::Ready(res) => {
+                        if let Some(Err(e)) = res {
+                            tracing::error!(
+                                "Error reading bytes from {}: {}, attempting to resume download",
+                                this.http_url,
+                                e
+                            );
+                            tracing::debug!("{:?}", e);
+                        }
+                        if *this.cursor_pos == *this.total_bytes {
+                            return Poll::Ready(Ok(()));
+                        }
+                        continue;
```

Comment:
i think you can remove this in a enum variant, but i like it explicit anyway.

Reply:
What do you mean?